### PR TITLE
Linked list.

### PR DESCRIPTION
**Pull request by TomaszRewak**
Sorry for using raw pointers, but creating bidirectional linked list with smart pointers seemed somehow wrong to me.

### CODE REVIEW

File: course02/homework01/Tomasz/list.h [hunk 1]
```diff
@@ -0,0 +1,480 @@
...
+		Node* next;
+		Node* prev;
+	};
+
+public:
+	template<typename E>
+	class Iterator
+	{
+	private:
+		Iterator(const List<T>& owner, Node* node);
```
Comment (maro243) [inline course02/homework01/Tomasz/list.h]: Nice! I was thinking of doing an iterator but didn't have time

File: course02/homework01/Tomasz/list.h [hunk 2]
```diff
@@ -0,0 +1,480 @@
...
+		Node(const T& data);
+
+		T data;
+
+		Node* next;
+		Node* prev;
+	};
+
+public:
+	template<typename E>
```
Comment (maro243) [inline course02/homework01/Tomasz/list.h]: I'm curious why an iterator needs a template here?

Comment (TomaszRewak) [reply course02/homework01/Tomasz/list.h]: I'm using it to create const and mutable iterators without defining them separately. So the const iterator is described as `Iterator<const T>` and standard iterator is simply `Iterator<T>`. I probably might have also just used a `List<const T>::Iterator` as a const iterators and then I wouldn't have to define the template here. I don't know what would be the cleaner solution.

Comment (maro243) [reply course02/homework01/Tomasz/list.h]: Makes a lot of sense now, cheers!

Comment (TomaszRewak) [reply course02/homework01/Tomasz/list.h]: Np. 
I've just checked and:
- indeed if I use the same plain iterator for const and mutable purposes then I'm able to edit list elements even if the list itself is conts (which is undesirable),
- in fact I could not have used `List<const T>::Iterator` as the `owner` of the iterator is still simply `List<T>` and node is `List<T>::Node` so there is a constructor parameter mismatch.

Comment (david-grs) [reply course02/homework01/Tomasz/list.h]: Indeed; another point is that you need to be able to construct a const_iterator from an iterator (and not allow the conversion from const_iterator to iterator of course). You also want to provide more operations between the two kind of iterators, typically, operator==. 

In general the solution that is being used is to defined a templated iterator_base class, with iterator and const_iterator that inherit from it. 

Last but not least you didnt implement the full Iterator requirement, typically iterator_category, reference, etc are missing, so you wouldn't be able to use such iterator with STL algos. See https://en.cppreference.com/w/cpp/named_req/Iterator My advice would be to use iterator helper from boost/operators.hpp

File: course02/homework01/Tomasz/list.h [hunk 3]
```diff
@@ -0,0 +1,484 @@
...
+		Iterator operator++(int);
+		Iterator& operator++();
+		Iterator operator--(int);
+		Iterator& operator--();
+
+		E& operator*() const;
+		E* operator->() const;
+
+		bool operator==(const Iterator& second);
+		bool operator!=(const Iterator& second);
```
Comment (david-grs) [inline course02/homework01/Tomasz/list.h]: const is missing

File: course02/homework01/Tomasz/list.h [hunk 4]
```diff
@@ -0,0 +1,484 @@
...
+	clear();
+}
+
+template<typename T>
+List<T>& List<T>::operator=(const List<T>& second)
+{
+	clear();
+
+	for (const auto& element : second)
+		push_back(element);
```
Comment (david-grs) [inline course02/homework01/Tomasz/list.h]: You should have used copy and swap here; if any allocation fail or exception is being thrown from one of T's constructor, you already cleared your list and can't provide a strong exception guarantee. It is also more logic. 

Simply pass the List by value to invoke the copy ctor, and do assign the pointer

**Review by david-grs — APPROVED** [review]
Neat implementation! I understand your concern around smart pointers, although I must say that an implementation with unique_ptr+raw pointer works pretty well too The only thing to be aware of is the destruction that can overflow the stack due to recursion.

File: course02/homework01/Tomasz/list.h (hunk 2 above)
Comment (TomaszRewak) [reply course02/homework01/Tomasz/list.h]: Thanks! I didn't know that there were such requirements for iterators to have iterator_category etc. Should the `value_type` of `const iterator` be `const T` or simply `T`, though? I've implemented it as T, but I'm not sure if that's correct.

I have also fixed the problem with performing operations between `const iterator` and `iterator` by deriving the second from the first.

Comment (TomaszRewak) [reply course02/homework01/Tomasz/list.h]: I've actually started implementing it with unique_pointers at first but the fact that I had to treat `prev` pointers and `next` pointers in nodes differently seemd to me as a code pollution. But I'm aware that this list is a rather specific example and overall it's a better idea to use smart pointers :)

File: course02/homework01/Tomasz/list.h [hunk 5]
```diff
@@ -18,41 +19,64 @@ class List
...
+		ConstIterator(const List<T>& owner, Node* node);
 
 	public:
-		Iterator operator++(int);
-		Iterator& operator++();
-		Iterator operator--(int);
-		Iterator& operator--();
+		using value_type = T;
+		using difference_type = std::ptrdiff_t;
+		using reference = T & ;
```
Comment (david-grs) [inline course02/homework01/Tomasz/list.h:30]: This should be `const T&`

File: course02/homework01/Tomasz/list.h [hunk 6]
```diff
@@ -18,41 +19,64 @@ class List
...
+		using pointer = T * ;
```
Comment (david-grs) [inline course02/homework01/Tomasz/list.h:31]: This should be `const T*`

File: course02/homework01/Tomasz/list.h [hunk 7]
```diff
@@ -18,41 +19,64 @@ class List
...
-		E& operator*() const;
-		E* operator->() const;
+		ConstIterator operator++(int);
+		ConstIterator& operator++();
+		ConstIterator operator--(int);
+		ConstIterator& operator--();
 
-		bool operator==(const Iterator& second);
-		bool operator!=(const Iterator& second);
+		const T& operator*() const;
```
Comment (david-grs) [inline course02/homework01/Tomasz/list.h:39]: It would make sense return `reference` here!

File: course02/homework01/Tomasz/list.h [hunk 8]
```diff
@@ -18,41 +19,64 @@ class List
...
+		const T* operator->() const;
```
Comment (david-grs) [inline course02/homework01/Tomasz/list.h:40]: and here `pointer`

File: course02/homework01/Tomasz/list.h [hunk 9]
```diff
@@ -18,41 +19,64 @@ class List
...
+
+	protected:
+		const List<T>* mOwner;
 		Node* mNode;
 
 		friend List<T>;
 	};
 
-	using value_type = T;
+	class MutableIterator : public ConstIterator
```
Comment (david-grs) [inline course02/homework01/Tomasz/list.h:52]: The pattern used in general to implement iterators is to have an `IteratorBase` templated class, and to inherit from it from Iterator and ConstIterator

File: course02/homework01/Tomasz/list.h [hunk 10]
```diff
@@ -18,41 +19,64 @@ class List
...
-	using MutableIterator = Iterator<T>;
-	using ConstIterator = Iterator<const T>;
+		using ConstIterator::operator*;
+		using ConstIterator::operator->;
+
+	protected:
+		friend List<T>;
+	};
+
+	using value_type = T;
```
Comment (david-grs) [inline course02/homework01/Tomasz/list.h:73]: Other type alias are missing , e.g iterator, const_iterator, pointer, etc

**Pull request merged**